src/iocp/mod.rs

```diff
@@ -0,0 +1,820 @@
+
+            match sources.entry(socket) {
+                Entry::Vacant(v) => {
+                    v.insert(Pin::<Arc<_>>::clone(&socket_state));
+                }
+
+                Entry::Occupied(_) => {
+                    return Err(io::Error::from(io::ErrorKind::AlreadyExists));
+                }
+            }
+        }
+
+        // Update the packet.
+        self.update_packet(socket_state)
+    }
+
+    /// Update a source in the poller.
+    pub(super) fn modify(
+        &self,
+        socket: RawSocket,
+        interest: Event,
+        mode: PollMode,
+    ) -> io::Result<()> {
+        log::trace!(
+            "modify: handle={:?}, sock={}, ev={:?}",
+            self.port,
+            socket,
+            interest
+        );
+
+        // We don't support edge-triggered events.
+        if matches!(mode, PollMode::Edge) {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "edge-triggered events are not supported",
+            ));
+        }
+
+        // Get a reference to the source.
+        let source = {
+            let sources = self.sources.read().unwrap_or_else(|e| e.into_inner());
+
+            match sources.get(&socket) {
+                Some(s) => s.clone(),
+                None => {
+                    return Err(io::Error::from(io::ErrorKind::NotFound));
+                }
+            }
+        };
+
+        // Set the new event.
+        if source.as_ref().set_events(interest, mode) {
+            self.update_packet(source)?;
+        }
+
+        Ok(())
+    }
+
+    /// Delete a source from the poller.
+    pub(super) fn delete(&self, socket: RawSocket) -> io::Result<()> {
+        log::trace!("remove: handle={:?}, sock={}", self.port, socket);
+
+        // Get a reference to the source.
+        let source = {
+            let mut sources = self.sources.write().unwrap_or_else(|e| e.into_inner());
+
+            match sources.remove(&socket) {
+                Some(s) => s,
+                None => {
+                    // Just return.
+                    return Ok(());
+                }
+            }
+        };
+
+        // Indicate to the source that it is being deleted.
+        // This cancels any ongoing AFD_IOCTL_POLL operations.
+        source.begin_delete()
+    }
+
+    /// Wait for events.
+    pub(super) fn wait(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        log::trace!("wait: handle={:?}, timeout={:?}", self.port, timeout);
+
+        let deadline = timeout.and_then(|timeout| Instant::now().checked_add(timeout));
+        let mut packets = self.packets.lock().unwrap_or_else(|e| e.into_inner());
+        let mut notified = false;
+        events.packets.clear();
+
+        loop {
+            let mut new_events = 0;
+
+            // Indicate that we are now polling.
+            debug_assert!(!self.polling.swap(true, Ordering::SeqCst));
```

Comment:
this won't be run during release builds, right?

Reply:
I'm not sure, I *think* the values are still consumed even on release mode? Either way, it's better to make it more explicit. Fixed!

Reply:
[`debug_assert!`](https://doc.rust-lang.org/core/macro.debug_assert.html) gets completely eliminated on release builds. *Beware*.